### ISSUE TITLE
verification image demo got ValueError: cannot reshape array of size 1090560 into shape (427,640,4)

### ISSUE THREAD

**Issue by bityigoss** [open-mmlab/mmdetection#4926]
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
3. The bug has not been fixed in the latest version.

**Describe the bug**
run the verification demo code, ValueError: cannot reshape array of size 1090560 into shape (427,640,4)

**Reproduction**

1. What command or script did you run?
![image](https://user-images.githubusercontent.com/6988886/114136701-b98ccf80-993d-11eb-996b-dd33c6ce16f5.png)

2. Did you make any modifications on the code or config? Did you understand what you have modified?
**change device='cuda:0' to device='cpu'**
3. What dataset did you use?

**macos, python3.7 mmcv-full==1.3.0 mmdet==2.11.0**

Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/4.2.1
sys.platform: darwin
Python: 3.7.10 (default, Feb 26 2021, 10:16:00) [Clang 10.0.0 ]
CUDA available: False
GCC: Apple clang version 12.0.0 (clang-1200.0.32.29)
PyTorch: 1.8.1
PyTorch compiling details: PyTorch built with:
  - GCC 4.2
  - C++ Version: 201402
  - clang 12.0.0
  - Intel(R) Math Kernel Library Version 2020.0.1 Product Build 20200208 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CXX_COMPILER=/Applications/Xcode-12.GM.seed.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -Wno-deprecated-declarations -DUSE_PTHREADPOOL -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-unused-private-field -Wno-missing-braces -Wno-c++14-extensions -Wno-constexpr-not-const, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.1, USE_CUDA=OFF, USE_CUDNN=OFF, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=OFF,

TorchVision: 0.9.1
OpenCV: 4.5.1
MMCV: 1.3.0
MMCV Compiler: clang 12.0.0
MMCV CUDA Compiler: not available
MMDetection: 2.11.0+5ebba9a

**Error traceback**
Traceback (most recent call last):
  File "demo/demo.py", line 24, in <module>
    text_color=(72, 101, 241))
  File "/Users/yiqi/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/mmdet/models/detectors/base.py", line 352, in show_result
    out_file=out_file)
  File "/Users/yiqi/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/mmdet/core/visualization/image.py", line 166, in imshow_det_bboxes
    img_rgba = buffer.reshape(height, width, 4)
ValueError: cannot reshape array of size 1090560 into shape (427,640,4)

**Comment by hhaAndroid** [comment]
Hi, I use `demo.py` and found that the result is correct. Did you modify the code?

**Comment by bityigoss** [comment]
![image](https://user-images.githubusercontent.com/6988886/115194407-0469cc80-a120-11eb-8ade-892a1cb927de.png)

sorry, here is my test code. I replace the test image with another one, it works. when use demo.jpg, the error: **ValueError: cannot reshape array of size 1090560 into shape (427,640,4)**,  btw, 1090560= 626 * 640 * 4

**Comment by EkAugust** [comment]
I tested 3 different size images, only 1 image could output the result correctly. The demo.jpg and the other one images outputted the same error. So I observed the image size information of the error.

you can see that
427 * 640 * 4 = 1,093,120
1093120 - 1090560 = 2560
2560 / 4 = 640.
So I guess the figure size is 426 * 640.

Anyway, let's debug this error. The full error message is shown below.

```
Traceback (most recent call last):
  File "main.py", line 17, in <module>
    show_result_pyplot(model, img, result)
  File "/usr/local/anaconda3/envs/openmmlab/lib/python3.7/site-packages/mmdet/apis/inference.py", line 240, in show_result_pyplot
    text_color=(72, 101, 241))
  File "/usr/local/anaconda3/envs/openmmlab/lib/python3.7/site-packages/mmdet/models/detectors/base.py", line 353, in show_result
    out_file=out_file)
  File "/usr/local/anaconda3/envs/openmmlab/lib/python3.7/site-packages/mmdet/core/visualization/image.py", line 172, in imshow_det_bboxes
    img_rgba = buffer.reshape(height, width, 4)
ValueError: cannot reshape array of size 1090560 into shape (427,640,4)
```

Firstly, I got into image.py and set a breakpoint at line 172. The result window and image was correctly shown. However, it was only an intermediate result. I observed the variables and found that the variable of line 119 is wrong.

```
 fig = plt.figure(win_name, frameon=False)  # fig: Figure (640*426)
```

The fig is initialized with size(1280\*960). Line 125:`fig.set_size_inches((width + EPS) / dpi, (height + EPS) / dpi)` changes the size into (640\*426). 

Then I found an annotation in line 123.

```
    # add a small EPS to avoid precision lost due to matplotlib's truncation
    # (https://github.com/matplotlib/matplotlib/issues/15363)
    fig.set_size_inches((width + EPS) / dpi, (height + EPS) / dpi)
```

EPS didn't work in macOS system and the same issue 15363 in matplotlib happened again.

**Comment by EkAugust** [comment]
I changed EPS from 0.01 to 0.1. It still doesn't work. Because dpi is set to 200 and we can see that
```
427/200*200
PyDev console: starting.
426.99999999999994
```
The truncation problem happens again. XD
I think you can solve it by yourself with my analysis.

**Comment by EkAugust** [comment]
The error occurs when height or width is singular. You can set reshape(height, -1, 4) or (-1, width, 4) to fix this error temporarily.

**Comment by levan92** [comment]
Hi, I can also confirm that using `buffer = np.frombuffer(stream, dtype='uint8')` with matplotlib's canvas stream, followed by reshaping back to image size often fails in macOS system. The same piece of code is able to run in Linux without any errors. It feels like certain bytes of the "stream" are dropping, therefore resulting in insufficient "pixels/sub-pixels" to reshape to the image size.

**Comment by monkeyDemon** [comment]
just like @EkAugust  says, modify reshape can fix this error temporarily.
and these lines maybe better
``` python
    if height % 2 != 0:
        img_rgba = buffer.reshape(-1, width, 4)
    else:
        img_rgba = buffer.reshape(height, -1, 4)
```

**Comment by levan92** [comment]
Probably more direct to check it this way. The issue is with the buffer stream not having enough values to shape up to the image shape + channels. reshape with -1 will still not work in some cases when it can't form the full cuboid with either width or height fixed. So, padding is more robust.
```python
if len(buffer)/width%4 == 0:
    img_rgba = buffer.reshape(-1, width, 4)
elif len(buffer)/height%4 == 0:
    img_rgba = buffer.reshape(height, -1, 4)
else:
    target_len = 4 * height * width
    assert len(buffer) < target_len
    buffer = np.pad(buffer, (0, target_len-len(buffer)))
    img_rgba = buffer.reshape(height, width, 4)
```
On the side note, based on the visualised output of those that require padding, those outputs all appear _slanted_. This means that the buffer stream is jumbled up, either due to `canvas.print_to_buffer()` or `np.frombuffer` in macOS?

**Comment by hhaAndroid** [comment]
> The error occurs when height or width is singular. You can set reshape(height, -1, 4) or (-1, width, 4) to fix this error temporarily.

Does this error only appear in macos?

**Comment by TangChiaHsin** [comment]
> > The error occurs when height or width is singular. You can set reshape(height, -1, 4) or (-1, width, 4) to fix this error temporarily.
> 
> Does this error only appear in macos?

The error can be traced back to matplotlib. Some developers said similar issues happened in Linux system. But when using mmdetection, the error only happened in macOS.

**Comment by ZwwWayne** [comment]
Resolved in #8434 and will be fixed in v2.25.1
